dali/python/nvidia/dali/external_source.py

```diff
@@ -102,7 +116,34 @@ def reset_indices(self):
         self.current_iter = 0
         self.current_sample = 0
 
+    def schedule_batch(self, pool, context_i, batch_size):
+        """Schedule computing new batch from source callback by the parallel pool."""
+        pool.schedule_batch(context_i, self.current_iter, [
+            self.callback_args(i) for i in range(batch_size)
+        ])
+        self.current_sample += batch_size
+        self.current_iter += 1
+
+    def schedule_and_feed(self, pipeline, pool, context_i, batch_size):
+        """Obtain the computed results of calling source callback in parallel pool and feed
+        the results to the ExternalSource nodes in `pipeline`.
+        Schedule the execution of the source callback in the pool to compute next batch.
+        Used by the parallel ExternalSource variant.
+
+        Args:
+            context_i (int): Index of the callback (in the list of parallel groups)"""
+        try:
+            callback_out = pool.receive_batch(context_i)
+            self.schedule_batch(pool, context_i, batch_size)
+            self.feed(pipeline, callback_out, batch_size)
+        except StopIteration:
```

Comment:
This is a race condition:
1. receive_batch just receives a shared memory object
2. schedule batch tells workers that they can work on a new batch
3. if `prefetch_queue_depth` is 1, the worker may thrash the shared memory chunk before we've fed it to the pipeline.
This is further aggravated when the external source works as no-copy.